src/api/crypto/Hash.cpp

```diff
@@ -90,6 +98,127 @@ auto Hash::bitcoin_hash_160(
+        const auto success =
+            uncompress_pubkey(data, writer(uncompressed_pubkey_temp));
+        if (false == success) {
+
+            throw std::runtime_error{"failed to calculate uncompressed pubkey"};
+        }
+        ReadView uncompressed_pubkey_without_prefix_readview(
+            reinterpret_cast<const char*>(uncompressed_pubkey_temp.data()),
+            uncompressed_pubkey_temp.size());
+
+        auto keccak_temp =
+            space(Provider::HashSize(opentxs::crypto::HashType::Keccak256));
+
+        const auto rc = Digest(
+            opentxs::crypto::HashType::Keccak256,
+            uncompressed_pubkey_without_prefix_readview,
+            writer(keccak_temp));
+
+        if (false == rc) {
+
+            throw std::runtime_error{
+                "failed to calculate intermediate keccak hash"};
+        }
+
+        return ethereum_address(reader(keccak_temp), destination);
+    } catch (const std::exception& e) {
+        LogError()(OT_PRETTY_CLASS())(e.what()).Flush();
+
+        return false;
+    }
+}
+
+auto Hash::uncompress_pubkey(const ReadView data, const AllocateOutput output)
+    const noexcept -> bool
+{
+    try {
+        if (false == output.operator bool()) {
+            throw std::runtime_error{"invalid output"};
+        }
+
+        if (data.size() > std::numeric_limits<int>::max()) {
+            throw std::runtime_error{"input too large"};
+        }
+
+        const auto uncompressed_without_prefix_size = 64;
+
+        auto buf = output(uncompressed_without_prefix_size);
+
+        if (false == buf.valid(uncompressed_without_prefix_size)) {
+            throw std::runtime_error{"failed to allocate space for output"};
+        }
+
+        auto hex_data = to_hex(
+            reinterpret_cast<const std::byte*>(data.data()), data.size());
+        ReadView hex_data_sv(hex_data.data(), hex_data.size());
+        auto uncompressed_pubkey_str =
+            opentxs::api::crypto::eliptic_curve::secp256k1::get_uncompressed_pubkey(hex_data_sv);
+
+        static const int eth_uncompressed_public_key_prefix_length = 2;
+        auto uncompressed_pubkey_str_without_04_prefix =
+            uncompressed_pubkey_str.substr(
+                eth_uncompressed_public_key_prefix_length);
+
+        // convert char to hex - treat 2 char sign as a byte
+        std::vector<char> uncompressed_pubkey_without_prefix_char_array;
+        boost::algorithm::unhex(
+            uncompressed_pubkey_str_without_04_prefix,
+            std::back_inserter(uncompressed_pubkey_without_prefix_char_array));
+
+        ReadView uncompressed_pubkey_without_prefix_ReadView(
+            uncompressed_pubkey_without_prefix_char_array.data(),
+            uncompressed_pubkey_without_prefix_char_array.size());
+        std::memcpy(
+            buf.as<unsigned char>(),
+            uncompressed_pubkey_without_prefix_ReadView.data(),
+            uncompressed_pubkey_without_prefix_ReadView.size());
```

Comment:
looks like equivalent of:
```
std::vector<unsigned char> uncompressed_key_bytes;
boost::algorithm::unhex(
    uncompressed_pubkey_str.substr(2),
    std::back_inserter(uncompressed_key_bytes);

    std::memcpy(
        buf.as<unsigned char>(),
        reinterpret_cast<char*>(uncompressed_key_bytes.data()), 
        uncompressed_key_bytes.size()
    );
```
i mean, i have a feeling it could be much shorter. at the end of the day what we need to do is:  
a) input -> hex  
b) hex -> lib -> uncompressed hex  
c) uncompressed hex -> output  

right?

Reply:
```
        static const int eth_uncompressed_public_key_prefix_length = 2;
        // convert char to hex - treat 2 char sign as a byte
        std::vector<unsigned char> uncompressed_pubkey_bytes;
        boost::algorithm::unhex(
            uncompressed_pubkey_str.substr(
                eth_uncompressed_public_key_prefix_length),
            std::back_inserter(uncompressed_pubkey_bytes));
        std::memcpy(
            buf.as<unsigned char>(),
            uncompressed_pubkey_bytes.data(),
            uncompressed_pubkey_bytes.size());
            ```

Reply:
That's my proposition - a combination of my and your code.
ReadView - was redundant
I like meaningful variables - even extra like "uncompressed_pubkey_str_without_04_prefix" but I know not everyone is a fan of it. I can get rid off it.
reinterpret_cast<char*> is not necessary - data() returns unsigned char*